libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Test.Cardano.Ledger.Core.Binary where
+
+import Cardano.Ledger.Core
+import Cardano.Ledger.MemoBytes (Memoized (RawType), zipMemoRawType)
+import Test.Cardano.Ledger.Binary.RoundTrip
+import Test.Cardano.Ledger.Binary.TreeDiff (ToExpr, expectExprEqual)
+import Test.Cardano.Ledger.Common
+
+specTxOutUpgrade ::
+  forall era.
+  ( EraTxOut (PreviousEra era)
+  , EraTxOut era
+  , Arbitrary (TxOut (PreviousEra era))
+  , HasCallStack
+  ) =>
+  Spec
+specTxOutUpgrade =
+  prop "upgradeTxOut is preserved through serialization" $ \prevTxOut -> do
+    case embedTrip (eraProtVerHigh @(PreviousEra era)) (eraProtVerLow @era) cborTrip prevTxOut of
+      Left err ->
+        expectationFailure $
+          "Expected to deserialize: =======================================================\n"
+            ++ show err
+      Right (curTxOut :: TxOut era) ->
+        curTxOut `shouldBe` upgradeTxOut prevTxOut
+
+specTxCertUpgrade ::
+  forall era.
+  ( EraTxCert (PreviousEra era)
+  , EraTxCert era
+  , Arbitrary (TxCert (PreviousEra era))
+  , HasCallStack
+  ) =>
+  Spec
+specTxCertUpgrade =
+  prop "upgradeTxCert is preserved through serialization" $ \prevTxCert -> do
+    case embedTrip (eraProtVerHigh @(PreviousEra era)) (eraProtVerLow @era) cborTrip prevTxCert of
+      Left err
+        | Right _ <- upgradeTxCert prevTxCert ->
+            -- We expect deserialization to succeed, when upgrade is possible
+            expectationFailure $
+              "Expected to deserialize: =======================================================\n"
+                ++ show err
+        | otherwise -> pure () -- Both upgrade and deserializer fail successfully
+      Right (curTxCert :: TxCert era)
+        | Right upgradedTxCert <- upgradeTxCert prevTxCert ->
+            curTxCert `shouldBe` upgradedTxCert
+        | otherwise -> expectationFailure "Expected upgradeTxCert to succeed"
+
+specTxAuxDataUpgrade ::
+  forall era t.
+  ( EraTxAuxData (PreviousEra era)
+  , EraTxAuxData era
+  , t era ~ TxAuxData era
+  , Memoized t
+  , Eq (RawType t era)
+  , ToExpr (RawType t era)
+  , Arbitrary (TxAuxData (PreviousEra era))
+  , HasCallStack
+  ) =>
+  Spec
+specTxAuxDataUpgrade =
+  prop "upgradeTxAuxData is preserved through serialization" $ \prevTxAuxData -> do
+    case embedTripAnn (eraProtVerHigh @(PreviousEra era)) (eraProtVerLow @era) prevTxAuxData of
+      Left err ->
+        expectationFailure $
+          "Expected to deserialize: =======================================================\n"
+            ++ show err
+      Right (curTxAuxData :: t era) ->
+        -- We need to do all this MemoBytes trickery because underlying bytes and thus the
+        -- equality of the same type will no longer be the same, despite that the value will
```

Comment:
This case will also be in `TxBody`. It is because `TxAuxData` is based on MemoBytes and equality on memo bytes also checks the equality of underlying bytes, which might not be the same between the two eras. That's why we need to use `zipMemoRawType` in order to compare just the underlying raw type, while ignoring the bytes